### PR TITLE
Parquet: clear metadata and project fields of ParquetRecordBatchStream::schema

### LINKED ISSUE #4023: `ParquetRecordBatchStream` Should Return the Projected Schema
**Describe the bug**

Let's say you're trying to async read a Parquet file on S3, and that file has metadata (like "created by"). There's an inconsistency:

`ParquetRecordBatchStream::schema` will produce a `Schema` object that includes that metadata.
But `ParquetRecordBatchStream` will yield `RecordBatch`es that have schema objects that don't have the metadata.

The problem is that if you create an  `ArrowWriter` using the first schema and then try to write batches from the stream to it, the schemas won't match (the writer is expecting metadata but each batch has a schema without metadata).

**Expected behavior**

I'd expect that either:
* `ParquetRecordBatchStream::schema` produces a `Schema` without metadata, or
* the `RecordBatch`es produced by `ParquetRecordBatchStream` have the exact same schema as what `::schema` returns, or
* `ArrowWriter` should tolerate its supplied schema differing from the batch schemas provided to `write()` in metadata

### PR DESCRIPTION
**Pull request by Jefffrey**
# Which issue does this PR close?

Closes #4023

# Rationale for this change
 

# What changes are included in this PR?

Ensure metadata hashmap of Schema from `ParquetRecordBatchStream::schema` is empty, to ensure exact same schema as for its RecordBatches it outputs.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
Whilst I agree with the spirit of this, stripping this metadata would be a major breaking change that would definitely break IOx and possibly some other workloads. Perhaps we could just document this potential disparity?

File: parquet/src/arrow/async_reader/mod.rs [hunk 1]
```diff
@@ -572,7 +572,15 @@ impl<T> std::fmt::Debug for ParquetRecordBatchStream<T> {
 }
 
 impl<T> ParquetRecordBatchStream<T> {
-    /// Returns the [`SchemaRef`] for this parquet file
+    /// Returns the [`SchemaRef`] for this parquet file.
+    ///
+    /// Note that unlike its synchronous counterpart [`ParquetRecordBatchReader`], the [`SchemaRef`]
+    /// returned here will contain the original metadata, whereas [`ParquetRecordBatchReader`]
+    /// strips this metadata.
```
Comment (tustvold) [inline parquet/src/arrow/async_reader/mod.rs:579]: Are you sure about this, they use the same logic? The difference is that the returned RecordBatch lack the metadata

Comment (Jefffrey) [reply parquet/src/arrow/async_reader/mod.rs:579]: Yes, see how schema is created for `ParquetRecordBatchReader` here:

https://github.com/apache/arrow-rs/blob/093a10e46203be1a0e94ae117854701bf58d4c79/parquet/src/arrow/arrow_reader/mod.rs#L609

And here:

https://github.com/apache/arrow-rs/blob/093a10e46203be1a0e94ae117854701bf58d4c79/parquet/src/arrow/arrow_reader/mod.rs#L622-L625

They only account for the fields, ignoring the metadata.

Comment (tustvold) [reply parquet/src/arrow/async_reader/mod.rs:579]: Oh this is the stream not the builder :facepalm: 

Yeah we should probably be returning the same schema as the RecordBatch we return, sorry for the confusion...

**Comment by tustvold** [comment]
Ok sorry for going back and forth on this, I hadn't quite grasped what the issue here was

```
#[tokio::test]
async fn test_projection_schema() {
    let mut metadata = HashMap::with_capacity(1);
    metadata.insert("key".to_string(), "value".to_string());

    let schema = Arc::new(
        Schema::new(Fields::from(vec![
            Field::new("a", DataType::Int32, true),
            Field::new("c", DataType::UInt64, true),
            Field::new("d", DataType::Float32, true),
        ]))
        .with_metadata(metadata.clone()),
    );

    let mut file = tempfile().unwrap();
    let mut writer = ArrowWriter::try_new(&mut file, schema.clone(), None).unwrap();
    writer
        .write(&RecordBatch::new_empty(schema.clone()))
        .unwrap();
    writer.close().unwrap();

    let builder = ParquetRecordBatchReaderBuilder::try_new(file.try_clone().unwrap()).unwrap();
    let sync_file_schema = builder.schema().clone();
    let mask = ProjectionMask::leaves(&builder.parquet_schema(), [1, 2]);
    let reader = builder.with_projection(mask).build().unwrap();

    let sync_reader_schema = reader.schema();

    assert_eq!(sync_file_schema.fields.len(), 3);
    assert_eq!(sync_file_schema.metadata, metadata);
    assert_eq!(sync_reader_schema.fields.len(), 2);
    assert_eq!(sync_reader_schema.metadata, HashMap::default());

    let file = tokio::fs::File::from(file);
    let builder = ParquetRecordBatchStreamBuilder::new(file).await.unwrap();
    let async_file_schema = builder.schema().clone();
    let mask = ProjectionMask::leaves(&builder.parquet_schema(), [1, 2]);
    let reader = builder.with_projection(mask).build().unwrap();

    let async_reader_schema = reader.schema();

    assert_eq!(async_file_schema.fields.len(), 3);
    assert_eq!(async_file_schema.metadata, metadata);
    assert_eq!(async_reader_schema.fields.len(), 2);
    assert_eq!(async_reader_schema.metadata, HashMap::default());
}
```

I think demonstrates the issue, in particular the schema returned by ParquetRecordBatchStream is incorrect, it should return the projected schema with the metadata removed.

**Comment by Jefffrey** [comment]
I'll revert back to the initial change, as well as account for projection which I wasn't aware it didn't account for either

**Comment by tustvold** [comment]
> I'll revert back to the initial change, as well as account for projection which I wasn't aware it didn't account for either

I think if you determine the schema from the ParquetField on ReaderFactory (which should be a DataType::StructArray) I think you might be able to get both in one

**Comment by Jefffrey** [comment]
Re-implemented the metadata stripping for `ParquetRecordBatchStream::schema` and also ensure it respects the projected columns

File: parquet/src/arrow/async_reader/mod.rs [hunk 2]
```diff
@@ -385,13 +385,28 @@ impl<T: AsyncFileReader + Send + 'static> ParquetRecordBatchStreamBuilder<T> {
             offset: self.offset,
         };
 
+        // Ensure schema of ParquetRecordBatchStream respects projection, and does
+        // not store metadata (same as for ParquetRecordBatchReader and emitted RecordBatches)
+        let projected_fields = match reader.fields.as_deref().map(|pf| &pf.arrow_type) {
```
Comment (tustvold) [inline parquet/src/arrow/async_reader/mod.rs:390]: This isn't quite correct in the case of a projection of a nested schema, its possible this isn't actually determined until the ArrayReader is constructed...

Comment (tustvold) [reply parquet/src/arrow/async_reader/mod.rs:390]: Although https://docs.rs/parquet/latest/parquet/arrow/fn.parquet_to_arrow_field_levels.html

takes a ProjectionMask and should apply it already?

Comment (Jefffrey) [reply parquet/src/arrow/async_reader/mod.rs:390]: I was a bit worried about this, as I couldn't find a straightforward way that the schema was constructed from ParquetField + ProjectionMask, as it seems done in the ArrayReader construction indeed.

Comment (tustvold) [reply parquet/src/arrow/async_reader/mod.rs:390]: I'll have a quick play, stand by

**Comment by Jefffrey** [comment]
I've reworked it to make use of the function introduced by #5149 

Hopefully I've used it correctly here

**Review by tustvold — APPROVED** [review]

**Pull request merged**